DOC: std/var: improve documentation of `ddof`

mdhaber opened the pull request:
gh-21670 reported that the documentation of `np.var` and `np.std` should include important terms like "sample variance"/"sample standard deviation" and "Bessel correction" to help users choose the appropriate `ddof` for their application. (It also reported a misunderstanding about how the "mean" was calculated, but this appears to have been ameliorated with the word "average" since the issue was opened.)

This PR goes a bit beyond that, giving a summary of factors that might go into the choice of `ddof`. 

Closes gh-21670. 

@steppi thought you might like to weigh in here.

mdhaber committed:
DOC: std/var: improve documentation of ddof

eendebakpt commented on numpy/_core/fromnumeric.py:
```diff
@@ -3755,15 +3776,33 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
+    unbiased estimator of the population variance. Other values of `ddof` may
+    have desirable properties for samples drawn from certain distributions;
+    for instance, if the population is normally distributed, ``ddof=1.5``
+    nearly eliminates the bias. Unfortunately, there is no simple, unbiased
```
The value 1.5 is more relevant for the standard deviation right? For the variance `ddof=1` is unbiased, so mentioning the desirable property that `ddof=1.5` nearly eliminates the bias seems odd.

mdhaber commented on numpy/_core/fromnumeric.py:
```diff
@@ -3755,15 +3776,33 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
+    of the variance when the population is normally distributed.
+
+    Use of ``ddof=1`` is known as the "Bessel correction", and the resulting
+    "corrected sample variance" (or simple *the* "sample variance") is an
+    unbiased estimator of the population variance. Other values of `ddof` may
+    have desirable properties for samples drawn from certain distributions;
+    for instance, if the population is normally distributed, ``ddof=1.5``
+    nearly eliminates the bias. Unfortunately, there is no simple, unbiased
+    estimator of the population variance that is valid for all distributions,
+    and reducing the bias often *increases* the mean squared error between the
+    sample variance and population variance. Consequently, the "best" choice of
+    `ddof` depends on the use case and is often subjective.
```
@eendebakpt How about simplifying it to:

```suggestion
    "corrected sample variance" (or simply *the* "sample variance") is an
    unbiased estimator of the population variance. However,
    reducing the bias often *increases* the mean squared error between the
    sample variance and population variance. Consequently, the "best" choice of
    `ddof` depends on the use case and is often subjective.
```

mdhaber commented:
One thing I thought about adding is that all this assumes that the mean is estimated as the sample mean. When the population mean is known and provided via `mean`, some things are different. In that case, `ddof=0` is still the MLE estimator for the normal distribution, but the sample variance is unbiased with `ddof=0`, I am not sure for the standard deviation, and I don't know anything about how that affects the mean squared error. Suggestions for addressing that?

rkern commented:
Probably nothing. The main issue is to explain what `ddof` *is* with enough conventional terms that people might already be familiar with so they will recognize what it represents (as it is very unconventional terminology that we invented). I don't think we should be comprehensive in explaining all possible ways that you might want to use it. Not in the docstring. A blogpost would probably be a better format.

Else you risk my dissertation on why `ddof=2` is the One Consistent Choice for both variance and stdev, and no one wants that. :wink:

mdhaber commented:
Well I do but you've hidden it well : ) Please send!

So should I cut the explanation and just list that "ddof=0" is "uncorrected" and "ddof=1" is "corrected" / Bessel Correction?

steppi commented:
> @steppi thought you might like to weigh in here.

I think it would be good to give the usual intuitive explanation for why a correction is needed. That the sample mean is the value $\bar{x}$ which minimizes the sum of squares $\sum_i (x_i - \bar{x})$, so replacing the population mean with the sample mean will naturally cause an underestimation of the variance. Something connecting the name degrees of freedom with the number of estimated parameters could be helpful too I think.

@rkern I think I've seen the term denominator degrees of freedom in books that predate NumPy in the context of estimating the standard deviation of the residuals of a regression model. I always thought it was a great naming choice. I'd be very interesting in hearing your thoughts on `ddof=2`.

mhvk commented:
@mdhaber - thanks for doing this! I agree with @steppi that perhaps most important is to mention why there is a bias if one estimates the mean as well - and otherwise I think it is fine to leave your text as you made it but add references to wikipedia
https://en.wikipedia.org/wiki/Variance#Sample_variance
https://en.wikipedia.org/wiki/Unbiased_estimation_of_standard_deviation

matthew-brett commented:
Thanks for doing this.   I think it would be very useful for add notes about the standard statistical terms here, which are:

* ddof=0 -> "population" variance / standard deviation
* ddof=1 -> "sample" variance / standard deviation

(for example see https://www.statology.org/sample-variance-vs-population-variance/).

These very common terms mean I think we have to be very careful in the using "sample standard deviation" in the explanation - I mean - we should never use that to mean anything other than the `ddof=1` value - and we probably shouldn't use it at all except to explain that statisticians use it to mean the `ddof=1` case.

Completely agreeing with @steppi - I am sure it would be worth some explanation of why using the mean of the values instead of the population mean, must result in a variance less than or equal to the population variance.

I personally like to stick to the terms "uncorrected" (ddof=0) and "Bessel-corrected" (ddof=1), while just noting that some people refer to the uncorrected as the "population" variance and the corrected as the "sample" variance.

It may well be worth starting off with something like:

> In statistics, we often consider the given set of values (here the array `a`) to be a *sample* from a larger *population*.  One use of the calculated variance or standard deviation of `a` is to *estimate* the variance or standard deviation for the population of which `a` is a sample.  For this use, it is often useful to correct for the fact that the simple and (for Numpy) default calculation using `ddof=0` gives a biased estimate for the population variance (and standard deviation).  This results from the use of `np.mean(a)` (call this $\bar{a}$) as an *estimate* of the population mean $\mu$.   Calculus shows us that $\bar{a}$ is the value that gives the smallest variance, of all possible values (Wikipedia).  Hence using $\bar{a}$ as the estimate for the mean in calculating the variance for `a` must give an equal or lower variance than we would get if using the population mean $\mu$.  We therefore say that the default `ddof=0` calculation is a *biased* (tends to be too low) estimate of the population variance. Call the `ddof=0` calculation the *uncorrected* variance.  By using calculus and expectations, it is possible to show that using `ddof=1` for the variance calculation gives an *unbiased* estimate for the population variance.   This is called the Bessel correction, and the resulting variance value is the *Bessel-corrected variance*.    Most confusingly, in statistics, it is common to refer to the uncorrected `ddof=0` calculation as the *population variance* of `a`, and the Bessel-corrected `ddof=1` result as the *sample variance* of `a`.

Or something to that effect.

matthew-brett commented:
Sorry - just to extend the justification for the statistics terminology - `ddof=0` is the population variance of `a` because it is precisely the variance of `a` where you consider `a` to be the entire population.  `ddof=1` is the sample variance of `a` if you consider `a` to be a sample from a larger population *and* you want an unbiased estimate of the population variance.  So "sample variance" in the sense of "best estimate of population variance given you assume `a` is a sample from a population".

rkern commented:
That's all reasonable, but I'd love to avoid _naming_ the `ddof=0` "uncorrected". It's a perfectly valid estimator that has certain properties (like being the maximum-likelihood estimator for Gaussians). Some people might like to correct _for_ those properties to get other properties, but it has its own virtues that and does not _need_ correction _a priori_.

When I said "dissertation", I meant it more in the colloquial sense of "overlong screed" than PhD dissertation. [I briefly outline my argument here in this old `scipy-user` thread](https://mail.python.org/pipermail/scipy-user/2008-December/019102.html). "Short" form is that by measuring the error of the variance estimate by arithmetic subtraction, per usual, you get the problems with the unbiased estimator for the standard deviation being different from the sqrt of the unbiased estimator of the variance. But because variances are positive scale quantities, the most natural way to measure error is geometrically in log-space, and that restores the equivalence across the sqrt. When you measure the error of the estimator, and thus bias, geometrically rather than arithmetically, the same `ddof` gets you the same minimum error. Unfortunately, it's `ddof=2`, which no one uses. And it's still distribution-dependent, so that's just the ideal Gaussian case. And none of this is appropriate for the docstring. :wink:

matthew-brett commented:
> That's all reasonable, but I'd love to avoid _naming_ the `ddof=0` "uncorrected". It's a perfectly valid estimator that has certain properties (like being the maximum-likelihood estimator for Gaussians).

Yes, that's totally fair - what would you suggest as the best name for the `ddof=0` option?

rkern commented:
It doesn't have a name, per se, not in the same way that `ddof=1` has the name "Bessel correction", which is the "Bessel correction" in any context. It's just the choice of `ddof` that you want in certain contexts; e.g. the MLE of the variance of a Gaussian, or treating the datapoints as an empirical distribution (i.e. a sum of Dirac deltas). I don't think it needs a name; I think you could just delete that sentence with no loss of comprehensibility as you never use the name again after defining it. Nor do I think it's a common term that people would recognize from prior experience like they might with "population variance" (which should be mentioned, as you do, but not as _the_ name for your good reasons).

For the most part, the goal here is to explain two things to people who are wondering why numpy is differing from pandas and their recollection of statistics:

* First, how to get numpy to behave like pandas and the rest of the universe.
* Second, assuage people that we're not _completely_ perverse for making `ddof=0` the default. Just mildly perverse.

matthew-brett commented:
Ok - then stick with `ddof=0` as the name, and avoid "uncorrected" - that seems reasonable.   For what it's worth, I like the `ddof=0` default.  I noticed that the Berkeley data science course uses the `ddof=0` calculation as - er - standard for teaching : https://inferentialthinking.com/chapters/14/2/Variability.html

mdhaber committed:
DOC: std/var: improve documentation of ddof, take 2

[skip actions] [skip azp] [skip cirrus]

mdhaber reviewed (commented):
I considered the feedback. Rather than add the intuitive understanding of bias, I removed  all the educational content about consistency, bias, and mean squared error, as "A blogpost would probably be a better format." 

My intent in opening this PR was to address gh-21670, which (essentially) was asking for "conventional terms that people might already be familiar with". For better or for worse, these include terms like "population"/"sample" and "uncorrected"/"corrected", so I've included those without commentary about whether they are good terms or not.

mdhaber committed:
DOC: std/var: make docstring raw to fix rendering

[skip actions] [skip azp] [skip cirrus]

mdhaber commented on numpy/_core/fromnumeric.py:
```diff
@@ -3755,15 +3773,35 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
+        var = d2.sum() / (N - ddof)  # note use of `ddof`
+
+    Different values of the argument `ddof` are useful in different
+    contexts. NumPy's default ``ddof=0`` corresponds with the expression::
```
```suggestion
    contexts. NumPy's default ``ddof=0`` corresponds with the expression:
```

mdhaber commented on numpy/_core/fromnumeric.py:
```diff
@@ -3755,15 +3773,35 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
+    and is sometimes (for better or for worse) called "population
+    variance", "uncorrected sample variance", or "biased sample variance".
+
+    Many other libraries define the variance of an array differently, e.g.::
```
```suggestion
    Many other libraries define the variance of an array differently, e.g.:
```

mdhaber committed:
DOC: std/var: remove colons to fix rendering

[skip actions] [skip azp] [skip cirrus]

mdhaber commented on numpy/_core/fromnumeric.py:
```diff
@@ -3588,24 +3588,42 @@ def std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
-    ``ddof=1``, it will not be an unbiased estimate of the standard deviation
-    per se.
+    The definition of the standard deviation of an array is not unique.
+    Assuming the input `a` is a one dimensional NumPy array and `mean` is
+    either provided as an argument or computed as ``a.mean()``, NumPy
```
```suggestion
    Assuming the input `a` is a one dimensional NumPy array and ``mean`` is
    computed as ``a.mean()`` if not provided, NumPy
```

mdhaber commented on numpy/_core/fromnumeric.py:
```diff
@@ -3755,15 +3773,35 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
-    ``ddof=0`` provides a maximum likelihood estimate of the variance for
-    normally distributed variables.
+    The definition of the variance of an array is not unique.
+    Assuming the input `a` is a one dimensional NumPy array and `mean` is
+    either provided as an argument or computed as ``a.mean()``, NumPy
```
```suggestion
    Assuming the input `a` is a one dimensional NumPy array and ``mean`` is
    computed as ``a.mean()`` if not provided, NumPy
```

matthew-brett commented on numpy/_core/fromnumeric.py:
```diff
@@ -3588,24 +3588,42 @@ def std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
-    function is the square root of the estimated variance, so even with
-    ``ddof=1``, it will not be an unbiased estimate of the standard deviation
-    per se.
+    The definition of the standard deviation of an array is not unique.
```
"not unique" sounds more technical than I think you mean.  How about "there are two standard variants of the standard deviation calculation" or something?

mdhaber replied:
I can use "there are several common variants", and that inclusion of the `ddof` term unifies them.

matthew-brett commented on numpy/_core/fromnumeric.py:
```diff
@@ -3588,24 +3588,42 @@ def std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
+    computes the standard deviation of an array as::
+
+        N = len(a)
+        d2 = abs(a - mean)**2
```
I guess `abs` is for the complex case?    If so - worth comment - as otherwise it is confusing.

mdhaber replied:
I guess. 
```suggestion
        d2 = abs(a - mean)**2  # abs for complex `a`
```

matthew-brett commented on numpy/_core/fromnumeric.py:
```diff
@@ -3588,24 +3588,42 @@ def std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
+
+        \sqrt{\frac{\sum_i{|a_i - \bar{a}|^2 }}{N}}
+
+    and is sometimes (for better or for worse) called "population
```
You do have some commentary here!  But I honestly think you should be explicit in explaining that a) this is the common term in the field of statistics in particular, and that b) it's based on the distinction between sample and population, and 'population' is in the sense where the values `a` are considered to be the whole population.   Is the term "uncorrected" common in fact?  Robert objected to that, I think on reasonable grounds.

mdhaber replied:
Thanks for catching that. I included that only for the reviewers and planned to remove it.

I know there was an objection to "uncorrected", and the rationale that the name shouldn't be used is reasonable. I included the word anyway because it it is a matter of opinion whether it should be used. However, it *is* used, and the purpose of this PR was to connect with what users already know rather than be prescriptive.

It is popular enough that it is the terminology I  use. I don't use "population", yet I included it first because it appears to be more common.

matthew-brett commented on numpy/_core/fromnumeric.py:
```diff
@@ -3588,24 +3588,42 @@ def std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
+        \sqrt{\frac{\sum_i{|a_i - \bar{a}|^2 }}{N - 1}}
+
+    The use of :math:`N-1` in the denominator is often called "Bessel's
+    correction", and the resulting quantity is sometimed called the "sample
```
Similarly here - I really do think you need some justification for the terminology - "sample" in the sense that we consider `a` to be sample from a population, and the variance calculation is biased (to be too low) as an estimate for the population, without the correction.

mdhaber replied:
I removed the educational comments. If I add back justification for "sample" and "population", I should also explain the other terms.

matthew-brett reviewed (commented):
Some comments, largely from the discussion in the thread.

mdhaber committed:
DOC: std/var: adjust per reviewer comments

[skip actions] [skip azp] [skip cirrus]

mdhaber committed:
DOC: std/var: additional tweaking

[skip actions] [skip cirrus] [skip azp]

mdhaber commented:
I took a stab at addressing your comments @matthew-brett. I removed the "uncorrected"/"corrected" terminology I was familiar with and motivated the "population"/"sample" terminology you preferred.

charris commented:
Thanks @mdhaber .
